### PR TITLE
Document atomicity of `write` and `pwrite`, and technical fix about POSIX

### PR DESCRIPTION
**Pull request by igrep**
See the commit message of each commit for details.

### CODE REVIEW

**Comment by programmerjake** [comment]
Is it technically possible to provide atomic writes with respect to non-WASI code, such as other processes? (Seems unlikely when a file is `mmap`ed.) If not, wording should be added to limit the scope of the atomicity to the WASI process (implementable with a mutex or similar) or some other suitable scope.

**Comment by igrep** [comment]
Thanks for pointing out. Reviewing [the related chapter of POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_09_07), it refers only to threads.
I'll fix.

**Comment by sunfishcode** [comment]
The related chapter in POSIX only provides these guarantees when the functions are operating on regular files or symbolic links. Unless we intend WASI to differ, we should preserve that aspect of the guarantee.

`write` etc. don't have a guarantee to write all the data or nothing. POSIX only says these "functions shall be atomic with respect to each other", which I read as saying that these functions' I/O won't be interleaved with other I/O, however they can presumably still produce incomplete results if they encounter errors such as running out of space on an output device. And in practice, some OS's also impose their own limitations -- on 64-bit Linux for example, calling `write` with a 4 GiB buffer (which is below `SSIZE_MAX`) to a normal file writes only 0x7ffff000 bytes.

**Comment by igrep** [comment]
> The related chapter in POSIX only provides these guarantees when the functions are operating on regular files or symbolic links

Thanks! I missed the phrase "when they operate on regular files or symbolic links".

> write etc. don't have a guarantee to write all the data or nothing. POSIX only says these "functions shall be atomic with respect to each other", which I read as saying that these functions' I/O won't be interleaved with other I/O, however they can presumably still produce incomplete results if they encounter errors such as running out of space on an output device

I agree. So I rewrote about what `atomicicity` stands for.

https://github.com/WebAssembly/WASI/pull/257/commits/13ab34bff433b422bbe267bfb78f6d776c45594c

File: phases/ephemeral/docs.md [hunk 1]
```diff
@@ -1190,7 +1190,12 @@ A buffer into which to write the preopened directory name.
 
 #### <a href="#pwrite" name="pwrite"></a> `pwrite(fd: fd, iovs: ciovec_array, offset: filesize) -> (errno, size)`
 Write to a file descriptor, without using and updating the file descriptor's offset.
-Note: This is similar to `pwritev` in POSIX.
+Note: This is similar to `pwritev` in Linux (and other Unix-es).
+
+Like Linux (and other Unix-es), [`pwrite`](#pwrite) should write all data in the `$fd` is
+for a regular file or a symbolic link to a regular file: Any calls of [`pwrite`](#pwrite)
```
Comment (sunfishcode) [inline phases/ephemeral/docs.md:1196]: Is `in` meant to be `if` here?

Also, POSIX doesn't say that `pwrite` etc. will write *all* the data; they could still be interrupted in the middle if an error occurs, such running out of space on the output device.

Also, as a minor correction, POSIX resolves symbolic links at `open` time, so it's not necessary to call out symblic links here in `pread` and `pwrite`. The reason POSIX mentions symbolic links in the paragraph linked above is that the list of affected functions includes `readlink` and `symlink` which operate on symbolic links themselves.

File: phases/ephemeral/docs.md [hunk 2]
```diff
@@ -1464,7 +1464,11 @@ A buffer into which to write the preopened directory name.
 
 #### <a href="#pwrite" name="pwrite"></a> `pwrite(fd: fd, iovs: ciovec_array, offset: filesize) -> (errno, size)`
 Write to a file descriptor, without using and updating the file descriptor's offset.
-Note: This is similar to `pwritev` in POSIX.
+Note: This is similar to `pwritev` in Linux (and other Unix-es).
+
+Like Linux (and other Unix-es), Any calls of [`pwrite`](#pwrite) (and other
```
Comment (MaxDesiatov) [inline phases/ephemeral/docs.md:1469]:
```suggestion
Like Linux (and other Unix-es), any calls of [`pwrite`](#pwrite) (and other
```

File: phases/ephemeral/docs.md [hunk 3]
```diff
@@ -1611,6 +1615,10 @@ The current offset of the file descriptor, relative to the start of the file.
 Write to a file descriptor.
 Note: This is similar to `writev` in POSIX.
 
+Like POSIX, Any calls of [`write`](#write) (and other functions to read or write)
```
Comment (MaxDesiatov) [inline phases/ephemeral/docs.md:1618]:
```suggestion
Like POSIX, any calls of [`write`](#write) (and other functions to read or write)
```

File: phases/ephemeral/witx/wasi_ephemeral_fd.witx [hunk 4]
```diff
@@ -157,7 +157,11 @@
   )
 
   ;;; Write to a file descriptor, without using and updating the file descriptor's offset.
-  ;;; Note: This is similar to `pwritev` in POSIX.
+  ;;; Note: This is similar to `pwritev` in Linux (and other Unix-es).
+  ;;;
+  ;;; Like Linux (and other Unix-es), Any calls of `pwrite` (and other
```
Comment (MaxDesiatov) [inline phases/ephemeral/witx/wasi_ephemeral_fd.witx:162]:
```suggestion
  ;;; Like Linux (and other Unix-es), any calls of `pwrite` (and other
```

File: phases/ephemeral/witx/wasi_ephemeral_fd.witx [hunk 5]
```diff
@@ -250,6 +254,10 @@
 
   ;;; Write to a file descriptor.
   ;;; Note: This is similar to `writev` in POSIX.
+  ;;;
+  ;;; Like POSIX, Any calls of `write` (and other functions to read or write)
+  ;;; for a regular file by other threads in the WASI process should not be
+  ;;; interleaved while `write` is executed.
```
Comment (MaxDesiatov) [inline phases/ephemeral/witx/wasi_ephemeral_fd.witx:260]:
```suggestion
  ;;; Like POSIX, any calls of `write` (and other functions to read or write)
  ;;; for a regular file by other threads in the WASI process should not be
  ;;; interleaved while `write` is executed.
```

**Comment by igrep** [comment]
@MaxDesiatov Thank you for the review!

**Comment by sunfishcode** [comment]
This looks good to me, thanks for seeing this through!

**Pull request merged**